### ISSUE TITLE
Distance Measures

### ISSUE THREAD

**Issue by florian-huber** [ZDDduesseldorf/HandScanAI#4]
**Edit: Übersicht Image Similarity Approaches**

## Image Similarity
> Image similarity can be thought of as a numerical representation of how alike two images are in terms of their visual content. There are several dimensions along which images can be similar, such as color, shape, texture, and composition.  ([Quelle](https://medium.com/scrapehero/exploring-image-similarity-approaches-in-python-b8ca0a3ed5a3))

What are we comparing? What kind of similarity are we looking for? Pictures with similar motiv and composition, or pictures that show similar objects in different poses? (See [comment](https://github.com/ZDDduesseldorf/HandScanAI/issues/4#issuecomment-2420561869))

### Histogram-based

> Histograms capture the distribution of pixel values in an image. By comparing the histograms of two images, you can measure their similarity. ([Quelle](https://medium.com/scrapehero/exploring-image-similarity-approaches-in-python-b8ca0a3ed5a3))

Possible Usecase: Visual Similarity, (color, motiv, composition). --> Für uns eher nicht relevant, da diese Features eigentlich nur Nebenprodukte der Aufnahme sind und uns die semantischen Eigenschaften der Hand mehr interessieren

### Structual Similarity Index (SSIM)

> [...] assesses the structural similarity between two images. It considers luminance, contrast, and structure, giving a score between -1 (dissimilar) and 1 (identical). ([Quelle](https://medium.com/scrapehero/exploring-image-similarity-approaches-in-python-b8ca0a3ed5a3))

> Pattern Recognition: Since SSIM mimics aspects of human perception, it could be used for recognizing patterns. When faced with issues like image scaling, translation and rotation, the algorithm's authors claim that it is better to use CW-SSIM,[[21]](https://en.wikipedia.org/wiki/Structural_similarity_index_measure#cite_note-:Gao2011-21) which is insensitive to these variations and may be directly applied by template matching without using any training sample. Since data-driven pattern recognition approaches may produce better performance when a large amount of data is available for training, the authors suggest using CW-SSIM in data-driven approaches. ([Quelle](https://en.wikipedia.org/wiki/Structural_similarity_index_measure))

Possible Usecases: Pattern recognition for parts or the whole picture regarding textures or line patterns of the hand. --> Wahrscheinlich fehler-/ problemanfällig, eventuell als Unterstützungsmetrik geeignet

### Feature-based approaches

> These methods extract salient features from images, such as edges, corners, or key points. Techniques like Scale-Invariant Feature Transform (SIFT) and Speeded-Up Robust Features (SURF) identify distinctive points in images, which can then be compared across images. ([Quelle](https://medium.com/scrapehero/exploring-image-similarity-approaches-in-python-b8ca0a3ed5a3))

Betrifft #8, evtl in Kombination mit #1 

### Deep Learning-Based Approaches

> Using pre-trained convolutional neural networks (CNNs) like ResNet, VGG, and Inception, you can extract deep features from images. [...] The similarity between images can then be computed based on the cosine similarity or Euclidean distance of these feature vectors. To improve the accuracy, we can preprocess the images. ([Quelle](https://medium.com/scrapehero/exploring-image-similarity-approaches-in-python-b8ca0a3ed5a3))

Betrifft u.a. Similarity- / Distance-Maße für Embeddings, siehe [Kommentar](https://github.com/ZDDduesseldorf/HandScanAI/issues/4#issuecomment-2420561869). 

### TODO: Metriken für Usecase prüfen und testen

Standard für Embeddings
- Cosine Similarity 
  - deals well with sparse data
  - scale-invariant (just the angle between vectors is important, magnitudes are ignored)
  - better in describing approximate location of data in high dimensional space

Bekannteste Similarity- / Distance-Maße (Werden verglichen in [Paper](https://ieeexplore.ieee.org/document/10270776)):
- Minkowski-Distance (generalization of:)
  - Euclidean Distance 
  - Manhattan Distance
- Hamming Distance
- Jaccard coefficient
- SØrensen-Dice similarity
- Cosine Similarity (am häufigsten genutzt für Embeddings)

Sharma et al. [[50](https://link.springer.com/article/10.1007/s11042-019-7424-8#ref-CR50)] (identity verification with geometrical features of hands)
- L1 ( = Manhattan Distance)
- [Earth mover distance (Emd)](https://en.wikipedia.org/wiki/Earth_mover%27s_distance) (auch: Wasserstein distance, wirkt, als wäre das ebenfalls eher eine distance metric für noise/ rauschen/ Motiv und Komposition), 
- Euclidean, 
- Cosine,
- [Chi-square (chisq or v2)](https://www.geeksforgeeks.org/chi-square-distance-in-python/) 
- modified hausdorff distance (Mhd), [ref1](https://link.springer.com/content/pdf/10.1007/bfb0015093.pdf)

(TODO) Weitere distance metrics, die man sich genauer anschauen könnte, je nachdem, was unsere Features/ Embeddings sind:
- Mahalanobis Distance ([ref1](https://www.machinelearningplus.com/statistics/mahalanobis-distance/))
  - multivariate distance metric, 
  - measures distance between a point and a distribution, 
  - multivariate equivalent of the Euclidean distance, 
  - performs well on a highly imbalanced dataset 
- Chebyshev (Chessboard-metric, wird oft mit euclidian und manhattan distance verglichen [ref1](https://www.datacamp.com/tutorial/chebyshev-distance))
- Correlation
- Bray-Curtis

**Comment by sansyan-q** [comment]
Wird in #62 mitbearbeitet/ umgesetzt

**Comment by sansyan-q** [comment]
Close Issue, Thema wird ggf. im Rahmen von #123 nochmal aufgegriffen
